components/builder-originsrv/src/migrations/origin_packages.rs

```diff
@@ -880,5 +427,312 @@ pub fn migrate(migrator: &mut Migrator) -> SrvResult<()> {
+        r#"DROP FUNCTION IF EXISTS get_origin_packages_for_origin_v3(text, bigint, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"CREATE OR REPLACE FUNCTION get_origin_packages_for_origin_v5 (
+                    op_ident text,
+                    op_limit bigint,
+                    op_offset bigint,
+                    op_visibilities text
+                 ) RETURNS TABLE(total_count bigint, ident text) AS $$
+                    BEGIN
+                        RETURN QUERY SELECT COUNT(*) OVER () AS total_count, op.ident
+                          FROM origin_packages op
+                          WHERE op.ident LIKE (op_ident  || '%')
+                          AND op.visibility = ANY(STRING_TO_ARRAY(op_visibilities, ','))
+                          ORDER BY op.ident DESC
+                          LIMIT op_limit
+                          OFFSET op_offset;
+                        RETURN;
+                    END
+                    $$ LANGUAGE plpgsql STABLE"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS get_origin_packages_unique_for_origin_v1(text, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS get_origin_packages_unique_for_origin_v2(text, bigint, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"CREATE OR REPLACE FUNCTION get_origin_packages_unique_for_origin_v4 (
+                   op_origin text,
+                   op_limit bigint,
+                   op_offset bigint,
+                   op_visibilities text
+                 ) RETURNS TABLE(total_count bigint, name text) AS $$
+                    BEGIN
+                        RETURN QUERY SELECT COUNT(*) OVER () AS total_count, op.name
+                          FROM origins o INNER JOIN origin_packages op ON o.id = op.origin_id
+                          WHERE o.name = op_origin
+                          AND op.visibility = ANY(STRING_TO_ARRAY(op_visibilities, ','))
+                          GROUP BY op.name
+                          ORDER BY op.name ASC
+                          LIMIT op_limit
+                          OFFSET op_offset;
+                        RETURN;
+                    END
+                    $$ LANGUAGE plpgsql STABLE"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS search_all_origin_packages_dynamic_v1(text, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS search_all_origin_packages_dynamic_v2(text, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS search_all_origin_packages_dynamic_v3(text, bigint, bigint, bigint)"#,
+    )?;
+    migrator.migrate(
+        "originsrv",
+        r#"DROP FUNCTION IF EXISTS search_all_origin_packages_dynamic_v4(text, bigint)"#,
+    )?;
+    migrator.migrate("originsrv",
```

Comment:
Will this work when the shards are in different DB instances?

Reply:
I honestly don't know.  We'll have the same problems with these that we have with the rest of the queries in our db that span shards.  I wrote up an issue about it at the beginning of August.  https://github.com/habitat-sh/habitat/issues/2884  I chatted with @reset a bit about how this will work long term, and no, this isn't the long term solution, but it's what we need to do to make it work right now.

Reply:
We’ll need to create a search service at some point which keeps an index of this data. For now since all shards are on the Sam DB we mind as well take the hit